mava/components/training/model_updating.py

```diff
@@ -95,17 +99,31 @@ def on_training_utility_fns(self, trainer: SystemTrainer) -> None:
             None.
         """
 
-        # Initilaise running statisitics here
-
+        # Initilaise target values running statisitics here
         if self.config.normalize_target_values:
-            trainer.store.running_stats_fn = compute_running_mean_var_count
-            running_stats = jnp.array([0, 0, 1e-4])
-        else:
-            running_stats = jnp.array([0, 1, 1e-4])
-
-        trainer.store.stats = {}
-        for net_key in trainer.store.trainer_agent_net_keys.keys():
-            trainer.store.stats[net_key] = running_stats
+            trainer.store.target_running_stats_fn = compute_running_mean_var_count
+
+        trainer.store.target_stats = {}
+        for agent in trainer.store.trainer_agent_net_keys.keys():
+            trainer.store.target_stats[agent] = dict(
+                mean=jnp.array([0]), var=jnp.array([0]), count=jnp.array([1e-4])
+            )
+
+        # Initilaise observations running statisitics here
+        # We only need to modify the variance since this is actually intiailised
+        # in the training/trainer on_building_init_end callback.
+        if self.config.normalize_observations:
+            trainer.store.norm_obs_running_stats_fn = update_and_normalize_observations
+            for agent in trainer.store.trainer_agent_net_keys.keys():
+                obs_shape = len(
+                    trainer.store.obs_norm_params[constants.OBS_NORM_STATE_DICT_KEY][
+                        agent
+                    ]["var"]
+                )
+                for x in range(obs_shape):
+                    trainer.store.obs_norm_params[constants.OBS_NORM_STATE_DICT_KEY][
+                        agent
+                    ]["var"][x] = 0
```

Comment:
Just to sanity check. Why is the variance set to zero initially? Should this not be done in the builder?

Reply:
So in the builder we set it to 1 so that initial observations are not normalised. In the trainer we set them to 0 and these are updated correctly during the training step. Eventually when the parameters are updated by the trainer, the executors will fetch the correct values. To have a good approximation the initial value should be 0. But we can not normalise with a value of 0, so initially we just keep the variance in the executor to be 1. This is the same approach I saw in acme implementation.

Reply:
Ah okay, that sounds good.  One thing that I'm still not sure about is what happens when a checkpoint is loaded. The `on_training_utility_fns` gets executed after the parameter client gets the variable values from the checkpointer. Would this not mean that if we load from a checkpoint we would overwrite the variance with zero? Is this intended behaviour?

Reply:
No the values in the executors only change when we call [executor update](https://github.com/instadeepai/Mava/blob/develop/mava/environment_loop.py#L164). I believe this can only happen after we have done atleast a training step so at that time the parameters should be correct. I get the correct behaviour for the test I wrote in a single process.

Reply:
Yes, but the problem is when restore from a checkpoint. In that case we don't want zero variance right? But this function is called after we load from the checkpointer and therefore it resets the variance to zero. Or does your single process test check the checkpointer as well?

Reply:
Check the new changes. We don't need this anymore.

Reply:
I just saw your latest message. So everything is loaded from the builder now? That should solve the problem.

Reply:
Yes, the solution was to add an std parameter to the stats. Initialise it to 0. Keep the variance to 0 and when the variance is updated, update the std as the sqrt of the variance.

Reply:
Great. Thanks @ulricharmel :slightly_smiling_face: